broadinstitute/single_cell_portal_core
# Using traditional MTX ingest for AnnData processed expression (SCP-5124)

#### BACKGROUND & CHANGES
This update switches AnnData processed expression ingest to use extracted MTX files, instead of direct MongoDB inserts.  This has multiple advantages over the previous model, as it is faster and reuses existing/tested code.  A user can now upload a single AnnData file and extract all 3 primary data types: clustering, metadata, and processed expression.  Also, a user no longer receives the initial success email for the extraction job as it does not contain any actionable information, and could lead to confusion.  Now, users only receive emails for downstream, traditional parse jobs of extracted flat files.

In addition, this contains refactoring of all the ingest-related "parameter" classes that leverage the `Parameterizable` module, as well as some other method consolidations.  This was done to standardize interfaces across similar classes and to make adding new attributes easier in the future.

Note: there is code in place to launch subsampling jobs for `AnnData` files, but this work is blocked by SCP-5140.  The logic works as intended, but the resulting MongoDB index violations cause the jobs to fail and all data to be deleted.  Once these violations are dealt with, subsampling for `AnnData` files will operate the same as normal SCP-formatted flat files.

#### MANUAL TESTING
1. Boot as normal and start all services related to file parsing
2. Sign in and create a new study
3. Select the `AnnData` UX, and upload the file at `test/test_data/anndata_test.h5ad`
   * There is clustering available for `obsm_key_name: X_tsne`
4. After uploading the file, note in the logs that the initial `extract` param contains `processed_expression` along with `cluster` and `metadata`
5. Once the first job finishes, note that the parsing job for expression launches with the `mtx` parameters:
```
:actions:
  :commands:
  - python
  - ingest_pipeline.py
  - "--study-id"
  - 6451311b94ec8f59c754a6d7
  - "--study-file-id"
  - 646644d9b636f8cb7d4a6389
  - "--user-metrics-uuid"
  - 0fc4d8e5-6c4c-422e-94a3-c78801757202
  - ingest_expression
  - "--matrix-file"
  - gs://fc-38dbd25a-eea7-4273-849b-732b1e664249/_scp_internal/anndata_ingest/SCP91_646644d9b636f8cb7d4a6389/h5ad_frag.matrix.processed.mtx.gz
  - "--matrix-file-type"
  - mtx
  - "--gene-file"
  - gs://fc-38dbd25a-eea7-4273-849b-732b1e664249/_scp_internal/anndata_ingest/SCP91_646644d9b636f8cb7d4a6389/h5ad_frag.features.processed.tsv.gz
  - "--barcode-file"
  - gs://fc-38dbd25a-eea7-4273-849b-732b1e664249/_scp_internal/anndata_ingest/SCP91_646644d9b636f8cb7d4a6389/h5ad_frag.barcodes.processed.tsv.gz
  :flags: []
```
6. (Optional) Wait until all 3 jobs complete, and then load the study and confirm that visualizations are enabled, and you can query for genes like `Foo` and `Bar`:
![Screenshot 2023-05-18 at 5 00 30 PM](https://github.com/broadinstitute/single_cell_portal_core/assets/729968/579fb8cf-742f-4054-89fd-eadb6516beb8)
